File: problem_solving/dp/cpp/bs_188_buy_sell_stock_k_tx.cpp
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/

3-dim vector .. days / tx / holding or not
                tx is from 0 to k, (k+1) array size
[i=days][j=transactions][holding=0|1]
HOLDING =>  (same as yesterday) Or (buy today)
//dp[i][j][1] = max( dp[i-1][j][1], (dp[i-1][j][0]) - prices[i])  )

NOT HOLDING => (same as yesterday) Or (sold today today)    => when we sell we consider it as a tx completed
//dp[i][j][0] = max( dp[i-1][j][0], dp[i-1][j-1][1]) + prices[i]  )
*/

#include<vector>
#include<iostream>

int max_profit(int k, const std::vector<int>& prices) {
    int n = prices.size();
    if(n <=1 || k == 0) {
        return 0;
    }

    // Handle large k
    if (k >= n / 2) {
        int profit = 0;
        for (int i = 1; i < n; i++) {
            if (prices[i] > prices[i-1]) {
                profit += prices[i] - prices[i-1];
            }
        }
        return profit;
    }

    //initialize the vector
    std::vector<std::vector<std::vector<int>>> dp(n, std::vector<std::vector<int>>(k+1, std::vector<int>(2, 0) ));
    dp[0][0][0] = 0; // No transactions, not holdin
    for(int j = 0; j <= k; ++j) {
        //holding stock on day 0, for any number of tx, the net will be -price
        dp[0][j][1] = -prices[0];//holding on day 0
    }

    //note we are starting from i=1, i =0 is taken care of
    for(int i = 1; i < n;++i) {
        for(int j = 0; j <= k ; ++j) {
            //NOT HOLDING
            //dp[i][j][0] = dp[i-1][j][0];//do nothing
            //if (j > 0 && dp[i-1][j-1][1] != INT_MIN) {
            if (j > 0) {
                dp[i][j][0] = std::max( dp[i-1][j][0], dp[i-1][j-1][1] + prices[i]  );
            }
            //HOLDING
            dp[i][j][1] = std::max( dp[i-1][j][1], dp[i-1][j][0] - prices[i]  );
            //std::cout << i << "     " << j << "     " << dp[i][j][0] << "     " << dp[i][j][1] << "     "  << std::endl;
        }
    }
    return dp[n-1][k][0];
}

/*
space optimization
    Not holding: dp[i][j][0] = max(dp[i-1][j][0], dp[i-1][j-1][1] + prices[i]) (if j > 0).
    Holding: dp[i][j][1] = max(dp[i-1][j][1], dp[i-1][j][0] - prices[i]).

    since i depends only on i-1, we dont need 3-dim array
*/
int max_profit_space_optimization(int k, const std::vector<int>& prices) {
    int n = prices.size();
    if(n <=1 || k == 0) {
        return 0;
    }

    if (k >= n / 2) {
            int max_profit = 0;
            for (int i = 1; i < n; ++i) {
                if (prices[i] > prices[i - 1]) {
                    max_profit += prices[i] - prices[i - 1];
                }
            }
            return max_profit;
    }

    //initialization
    std::vector<std::vector<int>> dp(k+1, std::vector<int>(2, 0) );
    dp[0][0] = 0;
    for(int j = 0; j <= k; ++j) {
        dp[j][1] = -prices[0];
    }

    for(int i = 1; i<n;++i) {
        for(int j=0; j <= k ;++j) {
            //NOT HOLDING
            int prev_not_holding = dp[j][0];//THIS IS THE OPTIMIZATION... VALUE FROM PREVIOUS DAY
            //dp[j][0] = dp[j][0]; // Do nothing => do we need this
            if (j > 0 && dp[j-1][1] != INT_MIN) {
                dp[j][0] = std::max( dp[j][0], dp[j-1][1] + prices[i]  );
            }

            //HOLDING
            if (prev_not_holding != INT_MIN) {
                dp[j][1] = std::max( dp[j][1], prev_not_holding - prices[i]  );
            }
            //std::cout << i << "     " << j << "     " << dp[i][j][0] << "     " << dp[i][j][1] << "     "  << std::endl;
            std::cout << "Day " << i << ", j=" << j << ": dp[" << j << "][0]=" << dp[j][0] << ", dp[" << j << "][1]=" << dp[j][1] << std::endl;
        }
    }
    std::cout << "Final dp[" << k << "][0] = " << dp[k][0] << std::endl;
    return dp[k][0];
}



int main() {
    std::vector<int> prices = {2, 4, 1};
    int k = 2;
    int result = max_profit(k, prices);
    std::cout << "result: " << result << std::endl;
    //2
    result = max_profit_space_optimization(k, prices);
    std::cout << "result-optimization: " << result << std::endl;
    

    prices = {3, 2, 6, 5, 0, 3};
    k = 2;
    result = max_profit(k, prices);
    std::cout << "result: " << result << std::endl;
    //7
    result = max_profit_space_optimization(k, prices);
    std::cout << "result-optimization: " << result << std::endl;
    return 0;
}

File: problem_solving/dp/py/bs_188_buy_sell_stock_k_tx.py
'''
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/description/

atmost k transactions..

    options on ith day, 
        no action
        sell if you have the stock
        buy if you dont have the stock

    prices
    n_transactions
    state

    find max profit
'''

'''
https://www.youtube.com/watch?v=qSGdZfkQ8C8

    k=0     => no profit
    day = 0 => no profit

no transactions => profit from previous day
        M= profit from previous day = dp[i][j-1]  i=transactions, j=days
        N=profit from today's tx.
        dp[i][j] = max(M, N)

        For N, we will be selling & the buy should have occured on any of the days 0...j-1,   i-1 tx should have been completed
            max( price[j] - price[i] + dp[i-1][x])
        profit on jth day = max( all possible profits )
                A = dp[i-1][x] = profit till day x because of i-1 transactions.  Note Buy + Sell = 1 transaction
                B = p[j] -p[x] = profit because of ith transaction
                total profit = A + B
                take max of (this for all the days.., No transactions  )
'''

#t:O(kN)
#s:O(kN)
def max_profit_iterative_effective_buy_price( k:int, prices:list[int]):
    N=len(prices)
    dp = [[0]*(N) for i in range(k+1)]
    print("dp: " + str(dp))
    if N < 2:
        return 0 #since we have atmost 1 price
    #if k < 1:
    #    return 0 #we have no transaction
    for i in range(1, k+1):
        effective_buy_price = prices[0]
        for j in range(1,N):
            #max of no-tx & sell
            #dp[i][j-1] = profit with no tx
            #helper(i, j, prices, dp) = profit with tx
            #effective buy prices = price - profit = prices[j]-dp[k-1][j]
            #dp[i-1][j] = profits from all previous transactions
            dp[i][j] = max(dp[i][j-1], prices[j] - effective_buy_price)  
            effective_buy_price = min(effective_buy_price, prices[j] - dp[i-1][j])
    print("dp: " + str(dp))
    return dp[k][N-1]


def max_profit_iterative( k:int, prices:list[int]):
    N=len(prices)
    dp = [[0]*(N+1) for i in range(k+1)]
    print("dp: " + str(dp))
    if N < 2:
        return 0 #since we have atmost 1 price
    if k < 1:
        return 0 #we have no transaction
    for i in range(1, k+1):
        for j in range(1,N+1):
            #max of no-tx & sell
            #dp[i][j-1] = profit with no tx
            #helper(i, j, prices, dp) = profit with tx
            #max of previous day profit, today's profit
            #dp[i][j-1] => previous day profit  #helper(i, j, prices, dp) => today' profit with transacations
            dp[i][j] = max(dp[i][j-1], helper(i, j, prices, dp))
    return dp[k][N]

# P - SP - BP
#this will give the max profit for selling.
#p[j] - p[x] + dp[k-1][x]  => assumes we are selling on jth day, with buy on xth day.  price diff j & i, max profit on xth day with k-1 tx
def helper(tx, p, prices, dp):
    val = 0
    for i in range(p):
        #compare the profits of buy on ith and sell on pth day + profit on ith day with tx-1
        #dp[tx-1][i] => profits due to tx-1 tx, till ith day,  prices[p] - prices[i] = profit due buy buy on ith and sell on pth day
        val = max(val, prices[p] - prices[i] + dp[tx-1][i])
    return val


prices = [0,1,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000,0,1000]
k=100
#result = max_profit_iterative(k, prices)
#print("result-iter: " + str(result))


result = max_profit_iterative_effective_buy_price(k, prices)
print("result-iter-ebp: " + str(result))

#-----------------------------

k = 2
prices = [2,4,1]
#Output: 2
#Explanation: Buy on day 1 (price = 2) and sell on day 2 (price = 4), profit = 4-2 = 2.

result = max_profit_iterative(k, prices)
print("result-iter: " + str(result))

result = max_profit_iterative_effective_buy_price(k, prices)
print("result-iter-ebp: " + str(result))

#-----------------------------


k = 2
prices = [3,2,6,5,0,3]
#Output: 7

result = max_profit_iterative(k, prices)
print("result-iter: " + str(result))

result = max_profit_iterative_effective_buy_price(k, prices)
print("result-iter-ebp: " + str(result))
